### PR TITLE
Feature/sink multiple input schemas

### PR DESCRIPTION
**Pull request by ibarbashov**
- Improve multiple input schemas handling in DB sink.
- Make MySQL autoReconnect property optional with default value.
- Add hook to support databases with case-sensitive identifiers.

### CODE REVIEW

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 1]
```diff
@@ -107,30 +108,36 @@ public void prepareRun(BatchSinkContext context) {
...
 
-    context.addOutput(Output.of(dbSinkConfig.referenceName,
-                                new DBOutputFormatProvider(dbSinkConfig, connectionString, dbColumns, driverClass)));
+    context.addOutput(Output.of(dbSinkConfig.referenceName, new DBOutputFormatProvider(
+      dbSinkConfig, connectionString, dbColumns, driverClass)));
   }
 
-  private void setColumnsInfo(List<Schema.Field> fields) {
+  /**
+   * Extracts column info from input schema. It is then used for metadata retrieval
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]:
```suggestion
   * Extracts column info from input schema. Later it is used for metadata retrieval
```

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 2]
```diff
@@ -107,30 +108,36 @@ public void prepareRun(BatchSinkContext context) {
...
+   * and insert query generation. Override this method if you need to escape column names
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]:
```suggestion
   * and insert during query generation. Override this method if you need to escape column names
```

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 3]
```diff
@@ -157,6 +164,8 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
            Statement statement = connection.createStatement();
            ResultSet rs = statement.executeQuery("SELECT * FROM " + dbSinkConfig.tableName + " WHERE 1 = 0")) {
         inferredFields.addAll(getSchemaReader().getSchemaFields(rs));
+      } catch (SQLException e) {
+        LOG.error("Error reading table metadata:", e);
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]:
```suggestion
        LOG.error("Error reading table metadata: ", e);
```

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 4]
```diff
@@ -157,6 +164,8 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
            Statement statement = connection.createStatement();
            ResultSet rs = statement.executeQuery("SELECT * FROM " + dbSinkConfig.tableName + " WHERE 1 = 0")) {
         inferredFields.addAll(getSchemaReader().getSchemaFields(rs));
+      } catch (SQLException e) {
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java:168]: Why do we only log exceptions and not fail?

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 5]
```diff
@@ -166,16 +175,15 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
...
 
   @Override
-  public void transform(StructuredRecord input, Emitter<KeyValue<DBRecord, NullWritable>> emitter) throws Exception {
+  public void transform(StructuredRecord input, Emitter<KeyValue<DBRecord, NullWritable>> emitter) {
     // Create StructuredRecord that only has the columns in this.columns
     List<Schema.Field> outputFields = new ArrayList<>();
-    for (String column : columns) {
-      Schema.Field field = input.getSchema().getField(column);
-      Preconditions.checkNotNull(field, "Column '%s' not found in an input record", column);
+    for (Schema.Field field: input.getSchema().getFields()) {
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]:
```suggestion
    for (Schema.Field field : input.getSchema().getFields()) {
```

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java [hunk 6]
```diff
@@ -166,16 +175,15 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
...
+      Preconditions.checkArgument(columns.contains(field.getName()), "Column not found for input field '%s'",
```
Comment (aielchiieva) [inline database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]: Please re-phrase:
```suggestion
      Preconditions.checkArgument(columns.contains(field.getName()), "Input field '%s' is not found in columns",
```

File: mysql-plugin/src/main/java/co/cask/mysql/MysqlAction.java [hunk 7]
```diff
@@ -48,7 +49,8 @@ public MysqlAction(MysqlActionConfig mysqlActionConfig) {
 
     @Name(MysqlConstants.AUTO_RECONNECT)
     @Description("Should the driver try to re-establish stale and/or dead connections")
-    public Boolean autoReconnect;
+    @Nullable
+    public Boolean autoReconnect = false;
```
Comment (aielchiieva) [inline mysql-plugin/src/main/java/co/cask/mysql/MysqlAction.java]: Why not use primitive type which is by default false? And this way you don't need to use nullable. Also please make sure you have default value set in the widget.

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java (hunk 1 above)
Comment (ibarbashov) [reply database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java (hunk 2 above)
Comment (ibarbashov) [reply database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java (hunk 4 above)
Comment (ibarbashov) [reply database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java:168]: exception added

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java (hunk 5 above)
Comment (ibarbashov) [reply database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java (hunk 6 above)
Comment (ibarbashov) [reply database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java]: done

File: mysql-plugin/src/main/java/co/cask/mysql/MysqlAction.java (hunk 7 above)
Comment (ibarbashov) [reply mysql-plugin/src/main/java/co/cask/mysql/MysqlAction.java]: done

**Comment by aielchiieva** [comment]
+1, please squash the commits before making external PR.

**Pull request merged**